### PR TITLE
Fix Geostationary projection boundary

### PR DESCRIPTION
**Pull request by dopplershift**
Finally put together enough spherical trigonometry to work out the proper boundary for the `Geostationary` projection.

Before:

before

After:

after

(See #1144 for sample plotting code.)

### CODE REVIEW

**Comment by QuLogic** [comment]
The proj implementation does some kind of swap of `x` and `y` depending on `sweep_axis`; do you need to do something similar?

File: lib/cartopy/tests/crs/test_geostationary.py [hunk 1]
```diff
@@ -54,19 +55,22 @@ def test_default(self):
...
         check_proj4_params(self.expected_proj_name, geos, other_args)
 
         assert_almost_equal(geos.boundary.bounds,
-                            (-8372.4040, -4171.5043, 8372.4040, 4171.5043),
+                            (-785616.1189, -785616.1189,
+                             785616.1189, 785616.1189),
+                            decimal=4)
+
+        # Checking that this isn't just a simple elliptical border
+        assert_almost_equal(np.array(geos.boundary.coords)[7],
```
Comment (QuLogic) [inline lib/cartopy/tests/crs/test_geostationary.py]: Do you really need `np.array` here?

Comment (dopplershift) [reply lib/cartopy/tests/crs/test_geostationary.py]: Apparently not. Removed.

**Comment by dopplershift** [comment]
The secret to the formula is some [spherical trigonometry](https://en.wikipedia.org/wiki/Spherical_trigonometry#Napier's_rules_for_right_spherical_triangles). It's based on solving a right triangle on the sphere where `c` or the hypotenuse is the radius (technically the angle subtended by the radius) in the original elliptical boundary. The legs of the triangle are related to our projected coordinates.

At a high enough satellite height, planar geometry works, but closer to the earth, we need to consider trig on the sphere. So we need to replace the `r * sin(theta)` and `r * cos(theta)` that generate the elliptical boundary with versions on the sphere (R2 and R6 at the link).

I've added a comment that at least points in this direction now.

**Comment by pelson** [comment]
How wonderful. Thanks for doing this @dopplershift. Great to see. 👍

**Pull request merged**